### PR TITLE
Add implementation of Rotary Embeddings

### PR DESCRIPTION
**Pull request by klei22**
# Rotary Position Embeddings

This is an implementation of standard RoPE based on modifcations of the repo's Symmetrical Rope implementation.

## Testing vs Existing Libraries

Checked this version of rotary embeddings against standard rope, output matches exactly for same inputs.

## Evaluations on Shakespeare with Char and Wikitext103 with Tiktoken

Sharing Image of current standings:
image

So far conventional RoPE has the most accurate next token prediction on Shakespeare_char and Wikitext103_Tiktoken.

The difference does get smaller for Wikitext103, suggesting the network with enough time can adapt to use the more memory efficient Symmetrical rope.

## Sweeps

From the sweeps, it appears for now that 512 does the best in terms of the Symmetrical Rope.

We need to test if we increase the max context length if the best number increases (e.g. does 512 do best with 1024 angles?) 

Or is 512 just an ideal angle that the network likes for separation with symmetrical rope approach?

## "Trimmed" Rope

Will try to patch this PR with options for "Trimming" the number of pairs which need to be rotated.

This will potentially save compute and memory for standard rope, and just compute time for symmetrical rope.

### CODE REVIEW

**Comment by klei22** [comment]
image

Added in a shortrope feature to Symmetrical Rope, and will be adding one next to the standard rope implementation

**Comment by klei22** [comment]
Both Embeddings are close in validation loss, maybe a more efficient rope implementation (with more training) could match performance of standard rope? Also curious if finetuning might be an effective route as well.

image

**Review by gkielian — APPROVED** [review]
Thanks so much! Looks good :+1:

**Comment by gkielian** [comment]
Looks good, noticed lots of possible experiments, wanted to jot down here before I forget:

- [ ] inference quality with shorter rope
- [ ] resume training with shorter rope (finetuning of greater efficiency), especially TTFT
- [ ] resume training with symmetrical rope vs standard rope
- [ ] also try training with rope shortening for symmetrical variation
